src/h2n_fetcher.erl

```diff
@@ -231,37 +231,45 @@ has_native_code_and_plugins(AppName, TempDirectory, TargetPath, AllApps) ->
                                         compressed]),
     {HasPortSpec, BuildPlugins0} = analyze_rebar_config(TempDirectory,
                                                         DirListing),
-    BuildPlugins = filter_out_unknown_plugins(AppName, BuildPlugins0, AllApps),
-    BuildTool  = get_build_tool(TempDirectory),
+    BuildPlugins = filter_out_bad_plugins(filter_out_unknown_plugins(AppName, BuildPlugins0, AllApps)),
+    BuildTool  = get_build_tools(DirListing),
+    io:format("Got ~p as build tool for ~s~n", [BuildTool, AppName]),
     {HasCSrc orelse HasPortSpec, BuildPlugins, BuildTool}.
 
--spec get_build_tool(file:name()) -> hex2nix:build_systems().
-get_build_tool(MetadataDirectory) ->
-    {ok, Items} = h2n_util:consult(filename:join(MetadataDirectory, "metadata.config")),
-    case lists:keysearch(<<"build_tools">>, 1, Items) of
-        {value, {_, BuildTools}} ->
-            resolve_build_tool(BuildTools);
-        false ->
-            rebar3
+-spec get_build_tools([string()]) -> hex2nix:build_systems().
+get_build_tools(DirListing) ->
+    case lists:foldl(fun resolve_build_tool/2, none, DirListing) of
```

Comment:
It's really tedious reading through resolve_build_tool/2 making sure that it implements precedence correctly. What if you did created a map or struct with fields like "has_rebar". So first you walk the files and update the struct. At that point you know what files you have, and you can then use pretty straight-forward logic for picking which thing you want to use